// boxes/lib/methods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// only the array methods that expose the target of the proxy
const methods = {
    every: (arr, proxy) => (fn) => arr.every((val, i) => fn(val, i, proxy)),
    filter: (arr, proxy) => (fn) => arr.filter((val, i) => fn(val, i, proxy)),
    find: (arr, proxy) => (fn) => arr.find((val, i) => fn(val, i, proxy)),
    findIndex: (arr, proxy) => (fn) => arr.findIndex((val, i) => fn(val, i, proxy)),
    flatMap: (arr, proxy) => (fn) => arr.flatMap((val, i) => fn(val, i, proxy)),
    forEach: (arr, proxy) => (fn) => arr.forEach((val, i) => fn(val, i, proxy)),
    map: (arr, proxy) => (fn) => arr.map((val, i) => fn(val, i, proxy)),
    reduce: (arr, proxy) => function (fn, init) {
        return '1' in arguments
            ? arr.reduce((val, i) => fn(val, i, proxy), init)
            : arr.reduce((val, i) => fn(val, i, proxy));
    },
    reduceRight: (arr, proxy) => function (fn, init) {
        return '1' in arguments
            ? arr.reduceRight((val, i) => fn(val, i, proxy), init)
            : arr.reduceRight((val, i) => fn(val, i, proxy));
    },
    some: (arr, proxy) => (fn) => arr.some((val, i) => fn(val, i, proxy))
};
exports.default = methods;
